querier.go

```diff
@@ -266,6 +267,62 @@ func (q *blockQuerier) Close() error {
 	return merr.Err()
 }
 
+// Bitmap used by func isRegexMetaCharacter to check whether a character needs to be escaped.
+var regexMetaCharacterBytes [16]byte
+
+// isRegexMetaCharacter reports whether byte b needs to be escaped.
+func isRegexMetaCharacter(b byte) bool {
+	return b < utf8.RuneSelf && regexMetaCharacterBytes[b%16]&(1<<(b/16)) != 0
+}
+
+func init() {
+	for _, b := range []byte(`.+*?()|[]{}^$`) {
+		regexMetaCharacterBytes[b%16] |= 1 << (b / 16)
+	}
+}
+
+func findSetMatches(pattern string) []string {
+	// Return empty matches if the wrapper from Prometheus is missing.
+	if len(pattern) < 6 || pattern[:4] != "^(?:" || pattern[len(pattern)-2:] != ")$" {
+		return nil
+	}
+	escaped := false
+	sets := []*strings.Builder{&strings.Builder{}}
+	for i := 4; i < len(pattern)-2; i++ {
+		if escaped {
+			switch {
+			case isRegexMetaCharacter(pattern[i]):
```

Comment:
Why is this check necessary?
Why not just allow every escaped character? (Or perhaps disallow `\0...` since that’s complicated).

Reply:
@brian-brazil Should I include the cases above?

Reply:
Sure, though check what Go does for escaping characters that don't need to be escaped.

Reply:
I'm sorry, I was thinking about the wrong language.
The set of escapes we care about here is documented at https://github.com/google/re2/wiki/Syntax and is painfully complicated.
There are lots of escapes we don't want to accept, like `\g`, `\p`, `\x`, so you do need something like what you have.

Reply:
I have rechecked, the special characters you meant like `\n`, `\a` are actually already included in the `else {}` part of `findSetMatches`. What I detect here are the special characters like `\\.`, `\\+` in regexp, which means after I find `\\`, I determine if the next char is special.